internal/ethapi: support block number or hash on state-related methods

ryanschneider opened the pull request:
This adds support for [EIP-1898](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1898.md) and supersedes https://github.com/ethereum/go-ethereum/pull/19459 which I forgot was based on 1.8.x.

### Outstanding Issues

- [x] Fix regression in GraphQL around `canonical` setting
- [ ] Potentially rename `BlockNumberOrHash` to `BlockSpecifier`
- [x] Determine behavior when specifier includes both hash and number (https://github.com/ethereum/EIPs/issues/2247)

### Rationale

As mentioned in #19394 there's a couple scenarios where I could see this being useful:

As a DApp developer, I might have logic like:

```golang
result := eth_call({....}, "latest")
// do some potentially slow processing w/ result
// send another RPC a couple seconds later
balance := eth_getBalance(0xsomeAdress, "latest")
```

The issue above is that `"latest"` could refer to totally different blocks.  So next one tries:

```golang
latest := eth_getBlockByNumber("latest", false)
result := eth_call({....}, latest.Number)
// do some potentially slow processing w/ result
// send another RPC a couple seconds later
balance := eth_getBalance(0xsomeAdress, latest.Number)
```

But even still, because of reorgs, block number 0x7654321 might refer to two different blocks depending on when it is used.  So my PR allows:

```golang
latest := eth_getBlockByNumber("latest", false)
result := eth_call({...}, latest.Hash)
// do some potentially slow processing w/ result
// send another RPC a couple seconds later
balance := eth_getBalance(0xsomeAdress, latest.Hash)
```

So now you  are sure you are referring to the same block in both calls.

This also becomes very useful for Exchanges or Infura-a-likes, basically anyone w/ enough RPC load or need for HA that they need to run more than one node, since the chances of 0x7654321 pointing to two different blocks on two different nodes is even higher.  And if one of the nodes hasn't seen that block hash yet, the caller gets an error rather than a 0-values result.  So, this actually skirts the breaking behavior of https://github.com/ethereum/go-ethereum/pull/18346 by continuing to behave like 1.8.x for block numbers, but returns explicit errors for block hashes.

In addition, to be compliant w/ EIP-1898 and the discussion in #19394 this PR accepts the hash as either a string or a `{blockHash:,}` object (w/ optional `requireCanonical` field, which defaults to false if not specified).

### GraphQL changes

The only real changes from #19459 are to make my changes work w/ the new GraphQL module, and actually they allowed for a fair bit of code cleanup in there as well, since it was going through some gymnastics to support looking up data by block number or hash.

However, two changes related to GraphQL (thus /cc @Arachnid and @kshinn):

- I removed all the canonical checks, since we now support looking up non-canonical blocks by hash.
- While I found it odd that a query like:

```graphql
block(hash: "0x....") {
   miner {
      balance
   }
}
```

Was returning the miner's balance at `"latest"`, I did *not* change this behavior since I think the best way to change this behavior requires schema changes and I feel like those should be addressed in a follow-on PR.

Arachnid commented on graphql/graphql.go:
```diff
@@ -1187,6 +1134,7 @@ func (s *Service) Start(server *p2p.Server) error {
 	var err error
 	s.handler, err = NewHandler(s.backend)
 	if err != nil {
+		log.Warn("whaaa?", "err", err)
```
Whaaa?

ryanschneider replied:
Hah, obviously some debugging code I didn't mean to commit. 😊

ryanschneider replied:
Force pushed 6bcd60f0f to remove this. :)

ryanschneider commented:
Hmm, @Arachnid one thing I realized w/ my change is that since the block param is just a scalar there's actually no way to specify `"latest"`.

And I think changing the GraphQL schema is probably outside the scope of this PR, so I think I'm going to revert that change and go back to a missing `block` argument defaulting to the latest block.

I'll start thinking of a follow-on PR to support the `block` param better but think it shouldn't be part of this PR.  Ideally we could use input unions but doesn't seem like they've been standardized in the GraphQL spec just yet.

ryanschneider commented:
Ok, force pushed an update that reverts the graphQL behavior back to the original, but leaves in place all the other refactors and adds a TODO about using graphQL union types as the input argument once those are supported.

As I mentioned above, I think we can make some schema-changing improvements to the graphQL layer once this PR is merged as we'll have access to the data via hash in a uniform way, but I don't want to complicate this PR w/ GraphQL schema changes.

Arachnid commented:
Not sure I understand the issue; if the earlier PR didn't change the schema, why couldn't it use the same solution for specifying the 'latest' block?

ryanschneider commented:
> Not sure I understand the issue; if the earlier PR didn't change the schema, why couldn't it use the same solution for specifying the 'latest' block?

While the schema itself didn't change, earlier versions of the PR changed the behavior of a `nil` `BlockNumberArgs` from `"latest"` to "the same block number that the query is returning", but upon thinking about it more that had two flaws:

- Previously, omitting the `block:` argument on a subfield like `miner` was how one specified latest, but since the `block:` argument was a Long and not a rpc.BlockNumber there was no way for a user to specify latest w/ this change.
- Furthermore, most users don't run archive nodes, so while from a consistency standpoint a query like:

```graphql
{
   block(number: 10000) { 
      miner {
         balance
      }
 }
}
```

Might be expected to return the miner's balance at block 10,000, most user's actually don't have the unpruned state to be able to lookup balance at block 10,000, they only have the last couple hundred blocks of unpruned state to lookup balance in.

And if a user really wants to know the balance at that block number, they can use GraphQL variables like so:

```graphql
query HistoricalMinerBalance($blockNumber = 10000) {
  block(number: $blockNumber) {
    hash
    number
    miner(block: $blockNumber) {
      address
      balance
    }
  }
}
```

And as noted in my TODO comment, once graphQL supports union types on input arguments `block:` could be updated to accept a hash, tag, or number as well.

Arachnid commented:
Thanks for clarifying. The standard here is, rather than a union, to have multiple optional arguments, of which at most one is provided.

ryanschneider commented:
@holiman @karalabe any thoughts on this PR?  I see there are some conflicts w/ master now that I can clean up but don't want spend the effort if there's no interest in merging.  Thanks!

ryanschneider commented:
@holiman @karalabe _any_ chance of this being included in the 1.9.x roadmap?  If not I'll close.

egalano commented:
Friendly ping. Adding support for this would be really useful for us for the reasons @ryanschneider pointed out in the description. Could someone from the geth team PTAL and comment on the approach or implementation.

holiman commented on rpc/types.go:
```diff
@@ -105,3 +107,97 @@ func (bn *BlockNumber) UnmarshalJSON(data []byte) error {
+		return nil
+	default:
+		if len(input) == 66 {
+			hash := common.Hash{}
```
Why not use `common.HexToHash` here?

ryanschneider replied:
I just looked and it doesn't look like `common.HextoHash` does any error checking, since this is user input I think we need to use an unmarshal function.  If there's a helper that does error checking I could use instead I'm happy to switch to it.

ryanschneider commented:
Thanks @holiman I'll incorporate your feedback and rebase against master as soon as I get some cycles (probably next week as I'm our on-call engineer this week).

Arachnid commented:
@ryanschneider Weren't you going to back out the GraphQL changes?

ryanschneider commented:
> @ryanschneider Weren't you going to back out the GraphQL changes?

Not all the changes, just this bit:

> And I think changing the GraphQL schema is probably outside the scope of this PR, so I think I'm going to revert that change and go back to a missing block argument defaulting to the latest block.

That change was done in  6bcd60f

> Ok, force pushed an update that reverts the graphQL behavior back to the original, but leaves in place all the other refactors and adds a TODO about using graphQL union types as the input argument once those are supported.

I should be able to rebase this PR against master this afternoon, and will address Martin's feedback as well, but I don't anticipate the GraphQL portions changing significantly from what's in the current PR.

ryanschneider commented:
@holiman thoughts on renaming `blockNrOrHash rpc.BlockNumberOrHash` to `blockSpecifier rpc.BlockSpecifier`?

ryanschneider commented on eth/api_backend.go:
```diff
@@ -87,6 +109,32 @@ func (b *EthAPIBackend) BlockByNumber(ctx context.Context, blockNr rpc.BlockNumb
 }
 
+func (b *EthAPIBackend) BlockByNumberOrHash(ctx context.Context, blockNrOrHash rpc.BlockNumberOrHash) (*types.Block, error) {
+	if blockNr, ok := blockNrOrHash.Number(); ok {
```
Good catch, currently the spec does not mention precedence or have any test cases for it.

 @charles-cooper care to update the spec regarding this scenario?  I see four options:

- Option 1: Block Number takes priority
- Option 2: Block Hash takes priority
- Option 3: If both blockNumber and blockHash are specified, return an error indicating these values are exclusive
- Option 4: If both blockNumber and blockHash are specified, enforce that both values are correct, otherwise return an error (i.e. lookup by hash, and return an error if returned header.Number != blockNumber).

ryanschneider replied:
I opened https://github.com/ethereum/EIPs/issues/2247 to solicit feedback on which option people prefer.

ryanschneider replied:
As Charles mentioned on the linked issue, probably makes sense to stay compatible with the behavior in EIP-234 (the one that added `blockHash` to eth_getLogs and friends), and that EIP says the number and hash arguments are mutually exclusive, so I'll update my code to Option 3.

ryanschneider replied:
I've updated `BlockNumberOrHash.UnmarshalJSON` to return an error if both number and hash are specified with an error very similar to the one returned by `FilterCriteria.UnmarshalJSON` when both are specified as part of an EIP-234 request.

ryanschneider commented:
Ok, I've got the branch rebased against master and cleaned up all of Martin's suggestions (except the pending ones with have outstanding questions about), I should be able to push a new version of this PR tomorrow after doing some more local testing.

ryanschneider commented:
Ok, @holiman I've pushed a new version that addresses all of your feedback except the behavior when both block number and hash is specified.

However I just realized that I left some artifacts of the `canonical` field when merging my GraphQL changes, so I need to push another update this afternoon.

ryanschneider commented on graphql/graphql.go:
```diff
@@ -456,26 +456,17 @@ func (b *Block) GasUsed(ctx context.Context) (hexutil.Uint64, error) {
-			canonical: isCanonical,
+			backend:      b.backend,
+			numberOrHash: &num,
+			hash:         b.header.ParentHash,
```
Actually, with my changes this canonical check is no longer needed, so I've removed those checks.

ryanschneider commented:
BTW, @Arachnid and @holiman if you prefer, I can totally remove the GraphQL changes for now and submit them as a separate PR.  They have some benefit (e.g. account state lookups are now supported on non-canonical blocks, see this gist: https://gist.github.com/ryanschneider/c6dd944e77fe586df2af38a690bbff91) but I understand that it adds a fair amount of cognitive load to this PR.

ryanschneider commented:
re-rebased against master to fix conflicts w/ https://github.com/ethereum/go-ethereum/pull/19639 (we both added GetCanonicalHash to the LES handlers).

holiman commented:
Ping @karalabe , PTAL
